### DISCUSSION TITLE
Error Accessing nodeSize on Undefined in Cursor Collaboration Extension

### DISCUSSION THREAD

**Discussion by mendesanna** [ueberdosis/tiptap#5377] [General Questions] [answered]
I'm encountering an error in my collaborative editing application using Tiptap with the Y.js real-time collaboration. The error message is TypeError: Cannot read properties of undefined (reading 'nodeSize'), and it occurs when I have around 5 active users and I focus the text area and refresh the page. After the refresh the text area breaks and I get this error 
`Uncaught TypeError: Cannot read properties of undefined (reading 'nodeSize')
    at relativePositionToAbsolutePosition (chunk-Y6AJG3WA.js?v=807bfd99:263:37)
    at chunk-Y6AJG3WA.js?v=807bfd99:1171:20
    at Map.forEach (<anonymous>)
    at createDecorations (chunk-Y6AJG3WA.js?v=807bfd99:1157:25)
    at Plugin.init (chunk-Y6AJG3WA.js?v=807bfd99:1215:14)
    at _EditorState.reconfigure (chunk-2ZLAYVCB.js?v=807bfd99:2580:75)
    at Editor2.createView (chunk-V54IQVDY.js?v=807bfd99:9365:33)
    at new Editor (chunk-V54IQVDY.js?v=807bfd99:9202:10)
    at new Editor2 (@tiptap_react.js?v=a3e2b2dd:3312:5)
    at @tiptap_react.js?v=a3e2b2dd:3375:20`
    If I dont have the text area focused doesn't happen or if I remove the CollaborationCursor extension this error doesn't happen.
    Does anyone have any idea what could it be?

**Comment by nperez0111** [comment]
Hm, so this is definitely coming from the collaboration cursor since it is talking about rendering decorations.

It seems to me to be an issue with how the collaboration is initialized. Could you show some code on how you initialized the collaboration plugin, collaboration cursor and the collab provider?

**Reply by mendesanna** [reply]
Sure this is the editor initialization:
```
  const { doc, provider } = useYjsContext();
 const xmlFragment = doc.getXmlFragment(documentSegmentId);
 const localState = provider?.awareness.getLocalState();
const editor = useEditor(
    {
      extensions: [
        StarterKit.configure({
          history: false, // Disables default history to use Collaboration's history management
        }),
        ListKeymap,
        FakeSelection,
        CustomPasteHandler,
        TextAlign.configure({
          types: ['paragraph', 'heading'],
        }),
        CustomListBehaviorExtension,
        Link.configure({
          autolink: false,
          validate: (href) => /^https?:\/\//.test(href),
        }),
        Underline,
        Collaboration.configure({
          fragment: xmlFragment,
        }),
        CollaborationCursor.configure({
          provider,
          // still need to add this here otherwise tiptap will set user to null
          user: {
            name: localState?.name || `${accountData?.firstName} ${accountData?.lastName}`,
            color: localState?.color || generateRandomColor(provider?.awareness?.clientID ?? 1),
            dataPointsPageId,
          },
        }),
      ],
      onUpdate: ({ editor: e }) => {
        const htmlContent = sanitizeHtml(e.getHTML(), htmlConfigs);
        if (htmlContent !== fieldRenderProps?.input.value) {
          fieldRenderProps?.input.onChange(e.isEmpty ? null : htmlContent);
        }
      },
    },
    [provider],
  );
  ```
  My provider initialization (some private code removed)
  ```
type YjsContextType = {
  doc: Y.Doc;
  provider: WebsocketProvider | null;
  isYjsSynced: boolean;
};

const WS_STATUS_EVENT = 'status';
const PERSISTENCE_SYNC = 'synced';
enum WebSocketStatus {
  CONNECTING = 'connecting',
  CONNECTED = 'connected',
  DISCONNECTED = 'disconnected',
  RECONNECTING = 'reconnecting',
}

type StatusEvent = {
  status: WebSocketStatus;
};

export const YjsContext = createContext<YjsContextType | null>(null);

type YjsProviderProps = {
  children: React.ReactNode;
};

const YjsProvider = ({ children }: YjsProviderProps) => {
  const doc = useRef<Y.Doc>(new Y.Doc({ gc: true }));
  const [provider, setProvider] = useState<WebsocketProvider | null>(null);
  const [isYjsSynced, setIsYjsSynced] = useState(false);

  useEffect(() => {
    if (!session?.idToken || !tenant || !companyData?._id || !config.api.yjsUrls) {
      return;
    }

    const newDoc = new Y.Doc({ gc: true });
    const persistence = new IndexeddbPersistence(room, newDoc);

    const handleSync = (event: StatusEvent) => {
      if (event.status === WebSocketStatus.CONNECTED) {
        setIsYjsSynced(true);
      }
    };

    let newProvider: WebsocketProvider | null = null;

    persistence.once(PERSISTENCE_SYNC, () => {
      doc.current = newDoc;
      newProvider = new WebsocketProvider(apiUrl, room, newDoc);
      setProvider(newProvider);

      newProvider.on(WS_STATUS_EVENT, handleSync);
    });

    return () => {
      newDoc.destroy();
      newProvider?.off(WS_STATUS_EVENT, handleSync);
      newProvider?.disconnect();
    };
  }, [companyData?._id, config.api.yjsUrls, session?.idToken, tenant]);

  const contextValue = useMemo(
    () => ({
      doc: doc.current,
      provider,
      isYjsSynced,
    }),
    [isYjsSynced, provider],
  );

  return <YjsContext.Provider value={contextValue}>{children}</YjsContext.Provider>;
};

export default YjsProvider;


  ```
  I think thats it, if you need more please let me know

**Reply by nperez0111** [reply]
This is likely to do with your yjs awareness not being available when the collaboration plugin is initialized. The provider needs to exist before the plugin can be created. You may need a guard to keep the editor from initializing 

My guess is that your error is specifically hitting an issue where your y prosemirror mapping is not set when you actually go to intialize the plugin: https://github.com/yjs/y-prosemirror/blob/f41a3e161448fe23e83a5ad69297274e4de7fb09/src/plugins/cursor-plugin.js#L105
https://github.com/yjs/y-prosemirror/blob/f41a3e161448fe23e83a5ad69297274e4de7fb09/src/plugins/cursor-plugin.js#L75

You can try initializing it beforehand: https://github.com/yjs/y-prosemirror/blob/ac5d5c5b8e3ada88c8a3cd641b096864aa345032/README.md#L23 and passing this into the collaboration plugin as ySyncOptions: https://github.com/ueberdosis/tiptap/blob/b941eea6daba09d48a5d18ccc1b9a1d84b2249dd/packages/extension-collaboration/src/collaboration.ts#L56


But in any case, this is somewhat outside of tiptap and firmly in the y-prosemirror binding

**Reply by mendesanna** [reply]
Hmm maybe I didn't understand but I'm only rendering the editor component where is this useEditor hook if the provider is connected and if I log the provider.awareness.getLocalState() in the parent component I can see my name, color, etc

**Reply by nperez0111** [reply]
Before the useEditor hook, log the provider. Is it initialized on the first render of the useEditor component?

**Reply by mendesanna** [reply]
Yes and this only happens when i more than 4 users online, if I have less this error is not showing

**Reply by mendesanna** [reply]
Also if I log provider.awareness.getStates() before the useEditor hook I can see all the users and cursor info

**Reply by nperez0111** [reply]
Hm that is strange, I'm sorry this is outside of my knowledge of y-prosemirror

**Reply by mendesanna** [reply]
Thank you for the help anyways 🙏  I found this similar [issue](https://github.com/yjs/y-prosemirror/issues/156) on y-prosemirror so waiting for their answer

**Reply by mendesanna** [reply]
Although if you could help me on something else, after I updated the version of tiptap from 2.4.0 to 2.5.5 i noticed the cursor label is always flickering and I figured out it was because the provider in the useEditor dependencies. But I need that dependency otherwise when the connection of the websocket closes and connects again the text area doesn't sync to the updated state.

**Reply by nperez0111** [reply]
useEditor should only ever re-render when that provider value actually changes. If you are seeing that provider value changing, but you didn't switch documents then you have a bug in your implementation.

If you see the value not changing, then there could be a bug with `useEditor`, but you can see here that we are passing your deps through completely unaltered: https://github.com/ueberdosis/tiptap/blob/70eebfdb0ccb8a0a5f84425a7dc17c75accde0fb/packages/react/src/useEditor.ts#L118

**Reply by nperez0111** [reply]
It could be initializing the editor twice, but I don't think that there is harm in that: https://github.com/ueberdosis/tiptap/blob/70eebfdb0ccb8a0a5f84425a7dc17c75accde0fb/packages/react/src/useEditor.ts#L109-L112

You can also look at the performance guide to use `shouldRerenderOnTransaction` to reduce re-renders: https://tiptap.dev/docs/guides/performance#page-title

**Reply by mendesanna** [reply]
I'll look into that, thank you so much for the help 🙏

**Reply by mendesanna** [reply]
Hello again
So I figured it's not a rerendering problem, I tested [here](https://codesandbox.io/p/sandbox/tiptap1-77vxx2) with a very simple code and it's still happening the flicker of the label and it's sending hundreds of awareness messages per second on the websocket.

**Reply by nperez0111** [reply]
Hi, I don't have access to the sandbox, could you make it public?

**Reply by mendesanna** [reply]
Ah sorry, I though I made it public. What about now?

**Reply by nperez0111** [reply]
Ah this is because you are trying to use a different provider (y-websocket) instead of the hocuspocusprovider or tiptapcollabprovider which are the only ones we support.
Providers aren't interchangeable so you can't just use another provider in there.

https://github.com/ueberdosis/tiptap/blob/b941eea6daba09d48a5d18ccc1b9a1d84b2249dd/packages/extension-collaboration-cursor/src/collaboration-cursor.ts#L10-L15

Please our docs for using hocuspocus and tiptap: https://tiptap.dev/docs/collaboration/getting-started/install#add-collaboration

**Reply by mendesanna** [reply]
ah ok but I changed it to TiptapCollabProvider and still same issue

**Reply by nperez0111** [reply]
I think I traced down the issue, the second param of useEditor seemed to be the problem: https://codesandbox.io/p/sandbox/tiptap1-forked-slx2zx?file=%2Fsrc%2FTiptapEditor.tsx%3A8%2C40

I think that the editor made 2 instances of the editor and would flicker between them because the hidden editor instance was never cleaned up.

**Reply by nperez0111** [reply, marked as answer]
I just made a new release https://github.com/ueberdosis/tiptap/releases/tag/v2.5.7

I updated the package.json in your example and it worked for me.


Sorry to send you down a bunch of rabbit holes. This was totally not obvious

**Reply by euvl** [reply]
Hey there, I've just tried 2.5.7 and still getting same issue.

Hope any of this blabber be useful:

It's failing at here: 
<img width="512" alt="Screenshot 2024-07-26 at 2 46 42 pm" src="https://github.com/user-attachments/assets/7b80ce53-c479-46ed-a6e7-5fa07546f342">

Where `mapping` seems to be empty and `mapping.get(contentType)` seems to not return an object that has `nodeSize`. 

<img width="724" alt="Screenshot 2024-07-26 at 2 47 46 pm" src="https://github.com/user-attachments/assets/b9115ed6-a61b-4e94-bc9c-55cbd7bf37d0">

---

My use-case is exactly same as described by the author of this issue, except 2 people is enough to make it break:

User 1 does something like this:

<img width="330" alt="Screenshot 2024-07-26 at 2 47 56 pm" src="https://github.com/user-attachments/assets/870baa60-6c28-415a-a87e-cfd81a8e4a65">

User 2 just opens a page.

**Reply by nperez0111** [reply]
With the same code? I just checked that?

**Reply by euvl** [reply]
Oh sorry, i thought you meant that [v2.5.7](https://github.com/ueberdosis/tiptap/releases/tag/v2.5.7) fixes the issue. 
I was using v2.5.6 so I've updated everything to v2.5.7 and tried - issue persisted.

<img width="485" alt="Screenshot 2024-07-26 at 3 17 52 pm" src="https://github.com/user-attachments/assets/d943bbf6-eccf-4f2f-941e-9db5fee9ef3c">

**Reply by nperez0111** [reply]
I did mean 2.5.7, it fixed the issue this the codesandbox above.

The issue was in @tiptap/react so you need to upgrade that too @euvl

**Reply by euvl** [reply]
Got you, let me try upgrading everything one more time and will report! Thanks 🙌

**Reply by mendesanna** [reply]
Hello @nperez0111 indeed that solved the problem. Thank you so much 🙏

**Reply by mendesanna** [reply]
Hello @euvl I do still have that bug. I've opened an issue [here](https://github.com/yjs/y-prosemirror/issues/162) on y-prosemirror as it seems the problem is on their side but still have no answer 😞 
Also brought it up on the [forum](https://discuss.yjs.dev/t/yjs-with-tiptap/2769) but nothing

**Reply by nperez0111** [reply]
Recently I noticed that y-prosemirror added [this initProsemirrorDoc to the readme](https://github.com/yjs/y-prosemirror/blob/master/README.md?plain=1#L23), I'm uncertain if it needs to be initialized like that but maybe that has something to do with it, since it is the one that creates the mapping (it was created before automatically but maybe it is no longer the case?)

**Reply by euvl** [reply]
Hey @mendesanna , @nperez0111 

Yeah still got the bug as well 😔.
You are right, seems to be coming from y-prosemirror. 

Interestingly it's happening only on construction. If i open 2 editors and then select some text (with both editors being unfocused initially) - it works just fine but if i refresh one - it crashes.

**Reply by nperez0111** [reply]
@euvl did you try initing y-prosemirror like I mentioned above? collaboration ext should allow arbitrary ySyncOptions https://github.com/ueberdosis/tiptap/blob/b941eea6daba09d48a5d18ccc1b9a1d84b2249dd/packages/extension-collaboration/src/collaboration.ts#L56

**Reply by euvl** [reply]
@nperez0111 I tried last night but got a little stuck. 

To run `initProseMirrorDoc(type, schema)` I need schema but then `schema` is created within `useEditor(...)` and I don't have it at the moment of running `initProseMirrorDoc`.

**Reply by nperez0111** [reply]
You could try doing `getSchema` (exported from `@tiptap/core`) with all the extensions on the editor. Just to see if it would work on not. If it does, then I can look into whether this needs to be fixed in tiptap or not

**Reply by euvl** [reply]
Hey @nperez0111 I was not able to make it fully work, but can confirm that calling `initProseMirrorDoc(type, schema)` creates proper mapping, which seems to be missing and causing an error with nodeSize above.

 Passing them to `ySyncOptions` didn't do the trick for me. Editor just doesn't render anything.
 
<img width="473" alt="Screenshot 2024-07-27 at 5 39 20 pm" src="https://github.com/user-attachments/assets/e4020b83-2788-41cf-ad95-9b517db8b84b">

**Reply by euvl** [reply]
@nperez0111 I've just tried downgrading y-prosemirror to ` "y-prosemirror": "1.2.6"` - version before introducing `mapping` (commit here for 1.2.7 - https://github.com/yjs/y-prosemirror/commit/7c78c19cab86355dfcfdd162f1428d58968c3901) and everything works without errors!

@mendesanna If you need a quick solution - downgrading to 1.2.6 seems to do the trick!

**Reply by nperez0111** [reply]
Hm, this tells me that we may need to do this initialization in the collaboration plugin. But a breaking change in a patch release?

**Reply by euvl** [reply]
> Hm, this tells me that we may need to do this initialization in the collaboration plugin.

I think you are right. 

> But a breaking change in a patch release?

I was also surprised but seems yeah, should've been a minor version bump.

**Reply by euvl** [reply]
By the way @nperez0111 while I have your attention: 

Seems these lines might have a separate issue: 

https://github.com/ueberdosis/tiptap/blob/44e4eca47a51df3fa12294115b34649a28d6e3cd/packages/extension-collaboration/src/collaboration.ts#L169-L172

`onFirstRender` is a function or undefined therefore { ...onFirstRender } results always in an empty object. 

I'm guessing should be something like:

```
    const ySyncPluginOptions: YSyncOpts = {
      ...(this.options.ySyncOptions ? { ...this.options.ySyncOptions } : {}),
      onFirstRender: this.options.onFirstRender,
    }
```

Unless I'm missing something. 

---

Tried this and it seems to print "TEST":
```
      Collaboration.configure({
        document: document,
        onFirstRender: {
          // @ts-ignore
          onFirstRender: () => {
            console.warn('TEST!!!!')
          }
        }
      }),
```

**Reply by nperez0111** [reply]
That is a good catch @euvl I will open a PR for this

**Reply by mendesanna** [reply]
@euvl Thanks a lot that indeed solved the problem 🙏 
@nperez0111 Will you look into this issue?

**Reply by nperez0111** [reply]
Yea I will

**Reply by nperez0111** [reply]
I can't reproduce the nodeSize error in the codesandbox, can either of you help me with that. This is the sandbox version I have right now (with 2.5.7 packages now): https://codesandbox.io/p/sandbox/tiptap1-forked-d3658v?file=%2Fsrc%2FTiptapEditor.tsx%3A21%2C12

**Reply by RafaZeero** [reply]
For me, the error happened when one of the editor had the cursor outside the position ( 0, 0 ). When another user reloads the page, then that error shows up.

**Reply by geekruchika** [reply]
Shall we reopen this issue? I'm getting similar issue?

**Reply by euvl** [reply]
Hey, i tried fiddling with codesandbox but was not able to recreate the issue either. I am using HocuspocusProvider though so wonder if thats why. Also I am not using xmlFragment like in codesamle but the document itself:

```
        Collaboration.configure({
          document
        }),

        CollaborationCursor.configure({
          provider,
          user,
          render(user) {
            const color = user.color || ''
            const cursor =` .... `
            const parser = new DOMParser()
            const dom = parser.parseFromString(cursor, 'text/html')

            return dom.querySelector('.collaboration-cursor__caret')!
          }
        })
```

**Reply by nperez0111** [reply]
Upgrade your y-prosemirror version it seems like this was resolved upstream

**Reply by RafaZeero** [reply]
@nperez0111 you're absolutely right, y-prosemirror just released a patch for that specific bug!

It's related to this [issue](https://github.com/yjs/y-prosemirror/issues/156#issuecomment-2245332566)
